### DISCUSSION TITLE
Mental Model for Configuration

### DISCUSSION THREAD

**Discussion by StevenXL** [LazyVim/LazyVim#3307] [General]
I was wondering how you all **think** about configuration in LazyVim.

For example, I had to turn off the LSP for Haskell, and I was trying to follow the directions provide by the nvim-lspconfig plugin:

https://github.com/neovim/nvim-lspconfig

Those directions said to do something like:

```plaintext
local lspconfig = require('lspconfig')
lspconfig.rust_analyzer.setup {
  -- Server-specific settings. See `:help lspconfig-setup`
  settings = {
    ['rust-analyzer'] = {},
  },
}
```
I read `lspconfig-setup` and came up with the following modifications to the above:

```plaintext
local lspconfig = require('lspconfig')
lspconfig.rust_analyzer.setup {
  -- Server-specific settings. See `:help lspconfig-setup`
  settings = {
    ['hls'] = {autostart = false},
  },
}
```

This isn't the correct way to do it. Instead, within the files loaded by LazyVim in the `plugins` directory, I created this file:

```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        hls = {
          autostart = false,
        },
      },
    },
  },
}
```

The thing that worked was different than the thing that the plugin was asking me to do. Luckily, for `nvim-lspconfig`, the `plugins/examples.lua` file helped a ton, but what about for the next plugin that does not have examples?

**Comment by dusty-phillips** [comment]
It can get complicated. Under the hood, Lazyvim is just a collection of prewritten config calls that accept opts. Lazyvim provides defaults for those opts and you can merge overriding opts in with those defaults.

90% of the time, that is enough. The remaining 10% you may have to resort to rewriting the config call from scratch because it doesn’t have a merging solution. That usually involves copying the existing config from the full spec section of the Lazyvim website and then modifying it.

I wrote about the basics in https://lazyvim-ambitious-devs.phillips.codes/course/chapter-5/ but unfortunately the advance section isn’t published yet.

**Reply by rubiin** [reply]
waiting for new chapters as well. I will reference the book on my [config's readme](https://github.com/rubiin/init.lua) as my colleagues are using my config

**Comment by dpetka2001** [comment]
That's the thing with most Neovim distros to be honest. They use levels of abstraction to provide an easier configuration according to each maintainer's experience/opinion, therefore they tend to be opinionated. Under the hood the same calls are being made. 

For what you're asking the corresponding abstraction is being made by the [setup function here](https://github.com/LazyVim/LazyVim/blob/0f6ff53ce336082869314db11e9dfa487cf83292/lua/lazyvim/plugins/lsp/init.lua#L186-L201). It takes the table defined in `opts.servers[server_name]` as `server_opts` and passes it to `require("lspconfig")[server].setup(server_opts)`, so essentially the same thing is done.

@dusty-phillips I saw in the chapter that you referenced that you do `local Util = require("lazyvim.util")`. That module is being assigned to the global `LazyVim` as you can see [here](https://github.com/LazyVim/LazyVim/blob/0f6ff53ce336082869314db11e9dfa487cf83292/lua/lazyvim/config/init.lua#L1), since not too long ago, so maybe you would like to update that in your book, as it's not necessary to import the module at the top of each file any more. You can directly do, for example `LazyVim.root` in your case to do the same thing. I understand that it's not possible to keep up with all the latest changes that happen in LazyVim and this may seem like nitpicking. I just thought I'd share my thoughts on this, nothing more.

**Reply by rubiin** [reply]
@dpetka2001 I was stuck on a situation where I need to use config function but I read somewhere that would cause my config function to overwrite the config setup doneby lazyVim for that module . Is there a way to extend the config of lazy. Opts are not an option on my case as its post config stuffs

**Reply by dpetka2001** [reply]
@rubiin Could you please show an example of what you want to do?

**Reply by rubiin** [reply]
I was doing this on my  cmp config function. But turns out that overrides everything lazy sets up by default

```
    -- `/` cmdline setup.
    cmp.setup.cmdline('/', {
      mapping = cmp.mapping.preset.cmdline(),
      sources = {
        { name = 'buffer' }
      }
    })
```

**Reply by rubiin** [reply]
Taking reference from here https://github.com/chrisgrieser/.config/blob/f4f1e3c135f0338ff416f883b4940e42298a63a7/nvim/lua/plugins/completion-and-snippets.lua#L166

**Reply by dpetka2001** [reply]
I've never used it myself, but looking from past discussions I seemed to just put the function in the `cmp-cmdline` spec itself when I tested it. So, something like 
```lua
return {
  {
    "hrsh7th/cmp-cmdline",
    config = function()
      local cmp = require("cmp")
      ---@diagnostic disable-next-line: missing-fields
      cmp.setup.cmdline(":", {
        mapping = cmp.mapping.preset.cmdline({
          ["<C-j>"] = {
            c = function(fallback)
              if cmp.visible() then
                cmp.select_next_item()
              else
                fallback()
              end
            end,
          },
        }),
        sources = cmp.config.sources({
          { name = "path" },
        }, {
          {
            name = "cmdline",
            option = {
              ignore_cmds = { "Man", "!" },
            },
          },
        }),
      })
      cmp.setup.cmdline("/", {
        mapping = cmp.mapping.preset.cmdline(),
        sources = {
          { name = "buffer" },
        },
      })
    end,
  },
}
```
and it shouldn't mess up with `nvim-cmp` config function.

**Reply by dpetka2001** [reply]
Turns out you can also put them in nvim-cmp `opts` if you define `opts` as a function. Something like
```lua
  {
    "nvim-cmp",
    event = "CmdlineEnter",
    dependencies = {
      "hrsh7th/cmp-cmdline",
    },
    opts = function()
      local cmp = require("cmp")
      cmp.setup.cmdline(":", {
        completion = {
          completeopt = "menu,menuone,noinsert,noselect",
        },
        mapping = cmp.mapping.preset.cmdline(),
        sources = cmp.config.sources({
          { name = "cmdline" },
        }),
      })

      cmp.setup.cmdline({ "/", "?" }, {
        completion = {
          completeopt = "menu,menuone,noinsert,noselect",
        },
        mapping = cmp.mapping.preset.cmdline(),
        sources = {
          { name = "buffer", max_item_count = 3, keyword_length = 2 },
        },
      })
    end,
  },
```
If you prefer it more that way.

**Reply by rubiin** [reply]
> Turns out you can also put them in nvim-cmp `opts` if you define `opts` as a function. Something like
> 
> ```lua
>   {
>     "nvim-cmp",
>     event = "CmdlineEnter",
>     dependencies = {
>       "hrsh7th/cmp-cmdline",
>     },
>     opts = function()
>       local cmp = require("cmp")
>       cmp.setup.cmdline(":", {
>         completion = {
>           completeopt = "menu,menuone,noinsert,noselect",
>         },
>         mapping = cmp.mapping.preset.cmdline(),
>         sources = cmp.config.sources({
>           { name = "cmdline" },
>         }),
>       })
> 
>       cmp.setup.cmdline({ "/", "?" }, {
>         completion = {
>           completeopt = "menu,menuone,noinsert,noselect",
>         },
>         mapping = cmp.mapping.preset.cmdline(),
>         sources = {
>           { name = "buffer", max_item_count = 3, keyword_length = 2 },
>         },
>       })
>     end,
>   },
> ```
> 
> If you prefer it more that way.

i tried this way earlier but the cmpletions never show. the first method works though

**Reply by dpetka2001** [reply]
I tried exactly this configuration and completions show just fine.

**Reply by dpetka2001** [reply]
Did you add the `event = "CmdlineEnter"`? Because by default `nvim-cmp` loads on `InsertEnter` and this event doesn't trigger on cmdline.
